File: LeetCode/30-Day-LeetCoding-Challenge-master/Diameter_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def get_height(self, root: TreeNode):
        if not root:
            return 0
            
        left_height = self.get_height(root.left)
        right_height = self.get_height(root.right)
        
        self.ans = max(self.ans, left_height + right_height)
        return 1 + max(left_height, right_height)
    
    
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.ans = 0
        self.get_height(root)
        return self.ans
        

File: LeetCode/30-Day-LeetCoding-Challenge-master/Diameter_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int get_height(TreeNode* root, int& ans) {
        if(root == NULL) return 0;
        
        int left_height = get_height(root->left, ans);
        int right_height = get_height(root->right, ans);
        
        ans = max(ans, left_height+right_height);
        return 
            1+max(left_height, right_height);
    }
    
    int diameterOfBinaryTree(TreeNode* root) {
        if(root == NULL) return 0;
            
        int ans = INT_MIN;
        int height = get_height(root, ans);
        return ans;
    }
};
